### PR TITLE
[https://nvbugs/6445494][infra] Upgrade Triton to 3.7.0 for Torch 2.12.0 compatibility

### PR DESCRIPTION
**Pull request by HandongLi-01**
## Description

Upgrade the PyTorch stack from Torch 2.11.0/Triton 3.6.0 to Torch 2.12.0/Triton 3.7.0.

The NVIDIA PyTorch 26.05 DLFW image supplies Torch `2.12.0a0+5aff3928d8.nv26.05`. Therefore, the package requirement is `torch>=2.12.0a0,<=2.13.0a0`, while public PyTorch installation paths use Torch 2.12.0.

This PR migrates TensorRT-LLM to the Triton 3.7 APIs and re-vendors `triton_kernels` from `v3.7.0`.

### Triton API migration

- Replace `triton_kernels.matmul_ogs` with `triton_kernels.matmul`.
- Migrate renamed matmul arguments and precision configuration fields.
- Replace removed routing wrappers with `TritonRoutingData`, `RaggedTensorMetadata`, and plain gather/scatter tensors.
- Explicitly combine scattered GEMM2 expert outputs, masking invalid expert-parallel routing slots.
- Update MXFP4 layout handling and aligned quantization behavior.
- Preserve paired MXFP4 value/scale layouts on Hopper and the natural strided path on Blackwell.

### Dependency and CI updates

- Require Torch `>=2.12.0a0,<=2.13.0a0`, supporting the NVIDIA DLFW 26.05 Torch build.
- Pin Triton to `3.7.0`.
- Re-vendor `triton_kernels` from Triton `v3.7.0`.
- Update public installation paths to Torch `2.12.0` and torchvision `0.27.0`.
- Remove the obsolete Jenkins Torch 2.11 pin workaround.
- Refresh the security dependency lock with Torch `2.12.0` and Triton `3.7.0`.
- Retain the repository-managed DLFW/`torch_skip` CI images; no PR-specific image build or promotion changes remain.

## Validation

- Pre-commit checks
- DCO sign-off verification
- Poetry lock validation
- Vendored Triton version verification
- Python and shell syntax checks
- Verified that `2.12.0a0+5aff3928d8.nv26.05` satisfies the new Torch requirement

## PR Checklist

- [x] PR title follows the required format
- [x] PR description explains the change
- [x] CI rerun pending

### CODE REVIEW

File: triton_kernels/roofline.py [hunk 1]
```diff
@@ -198,7 +206,7 @@ def plot_roofline(series, flops_dtype, out_path, max_tbps="memset", max_tflops="
         max_tbps = get_memset_tbps()
     if not isinstance(max_tflops, int):
         assert max_tflops == "cublas"
-        max_tflops = get_cublas_tflops(flops_dtype)
+        max_tflops = get_blas_tflops(flops_dtype)
```
[the remark this answers is not in the document]
Comment (HandongLi-01) [reply triton_kernels/roofline.py:209]: Confirmed this label is upstream behavior: it is unchanged in Triton v3.7.1 and current main. This PR vendors v3.7.0 verbatim, and triton_kernels/README.md forbids direct edits. I am not introducing a downstream-only divergence; a92e1d401d excludes the vendored snapshot from CodeRabbit review.

File: triton_kernels/testing.py [hunk 2]
```diff
@@ -198,3 +202,132 @@ def compute_actual_scale(x, dtype, per_batch_scale=False):
...
+    # always set one slice size to zero
+    probs = torch.ones(n_slices, device=device) / n_slices
+    if n_slices > 1:
+        probs[2] += probs[1]
+        probs[1] = 0.
+    assignments = torch.multinomial(probs, total_size, replacement=True)
+    counts = torch.bincount(assignments, minlength=n_slices).to(dtype)
+    assert counts.sum().item() == total_size
+    assert len(counts) == n_slices
+    return counts
```
[the remark this answers is not in the document]
Comment (HandongLi-01) [reply triton_kernels/testing.py:263]: Confirmed the n_slices == 2 issue is present unchanged in Triton v3.7.1 and current main. This PR vendors v3.7.0 verbatim, and triton_kernels/README.md forbids direct edits. I am not introducing a downstream-only divergence; a92e1d401d excludes the vendored snapshot from CodeRabbit review.

File: triton_kernels/distributed_details/mesh.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+        device: torch.device,
+    ) -> None:
+        if self._is_initialized:
+            return
+
+        BLOCK_N = 32
+        BLOCK_M = 32
+        n_bytes_topk = n_tokens_global * n_expts_act * 4  # topk logits (float32): pessimistic estimate
+        n_bytes_topk += n_tokens_global * n_expts_act * 2  # topk indx (int16)
+        cdiv = lambda x, y: (x + y - 1) // y
```
[the remark this answers is not in the document]
Comment (HandongLi-01) [reply triton_kernels/distributed_details/mesh.py:159]: Resolved in a92e1d401d by excluding triton_kernels/** from CodeRabbit review. The same vendored path was already excluded from Ruff in pyproject.toml, so no upstream source was modified.

**Review by Shixiaowei02 — APPROVED** [review]

**Review by dc3671 — APPROVED** [review]

**Review by EmmaQiaoCh — APPROVED** [review]

**Review by govind-ramnarayan — APPROVED** [review]

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py [hunk 4]
```diff
@@ -150,6 +152,58 @@ def _routing_clear_bitmatrix(Bitmatrix, stride_bm, stride_bn, shape_bn, cutoff,
...
+    n_expts_act = rdata.n_expts_act
+    assert gemm2_output.shape[-2] % n_expts_act == 0, (
+        f"expected [n_tokens * n_expts_act, hidden] rows, got "
+        f"{gemm2_output.shape[-2]} rows for n_expts_act={n_expts_act}")
+    n_tokens = gemm2_output.shape[-2] // n_expts_act
+    gemm2_output = gemm2_output.view(n_tokens, n_expts_act,
+                                     gemm2_output.shape[-1])
+    mask = rdata.scatter_valid_mask.view(n_tokens, n_expts_act,
+                                         1).expand_as(gemm2_output)
+    combined, _ = triton_reduce(gemm2_output, dim=1, mask=mask)
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py:208]: Minor perf note: this masked reduce always runs, but for `n_expts_act == 1` without EP pruning it is a full-tensor copy that the old `matmul_ogs` path avoided entirely (it fused the scatter and skipped finalize on CUDA when top-k == 1). If any top-1 model uses this backend, consider short-circuiting: have the router record whether pruning occurred (`ep > 1`), and return `gemm2_output` directly when `n_expts_act == 1` and no slots were pruned. Note the shortcut is only safe in that case — under EP pruning the unwritten rows are uninitialized (`torch.empty`) and must be masked, as done here. Fine as a follow-up.

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py [hunk 5]
```diff
@@ -228,21 +280,21 @@ def __call__(self,
...
-            src_indx=metadata.row_sorted_indx,
-            dst_indx=metadata.col_sorted_indx,
+            scatter_valid_mask=metadata.row_sorted_indx != -1,
         )
+        # For compute (expert-sorted) row i, `col_sorted_indx[i]` is the
+        # expanded (token-major) row `token * n_expts_act + k`. The gather
+        # index must address rows of the activations, hence the division; the
+        # scatter index addresses rows of the expanded output. This matches
+        # what `matmul_ogs` derived from its gather/scatter wrappers.
+        gather_indx = torch.div(metadata.col_sorted_indx,
```
Comment (brnguyen2) [inline tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py:305]: Worth one extra sentence in this comment: with EP pruning, `col_sorted_indx` contains `-1` for pruned compute rows, which trunc-divides to `0` here — so those rows gather token 0's activation and compute garbage. That is safe only because their scatter destination is also `-1`, which the matmul kernel drops (same behavior as the old in-kernel division), but the invariant is non-obvious to a future reader.

Comment (HandongLi-01) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py:305]: Done in `b75a679dad` — added the invariant right above the gather:

```python
# Under EP pruning `col_sorted_indx` is -1 for pruned compute rows,
# which truncates to gather index 0 -- those rows read token 0's
# activation and compute garbage. That is safe only because their
# scatter destination is also -1, which the matmul kernel drops (same
# behavior as the old in-kernel division); the garbage is never
# written out.
```

File: tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py (hunk 4 above)
Comment (HandongLi-01) [reply tensorrt_llm/_torch/modules/fused_moe/fused_moe_triton.py:208]: Agreed — tracking as a follow-up. Benchmarked the current flow on H200 to size it, and the data says this short-circuit matters more than I expected at high concurrency.

gpt-oss-120b MXFP4, TP4/EP4, ISL/OSL 1024/1024, both arms in the same Slurm allocation on one node. Note the stock `pytorch_model_config.py` never sets `moe_config.backend`, so a default `trtllm-bench` run doesn't reach this code at all — I forced `moe_config: {backend: TRITON}` via `--extra_llm_api_options` and verified `TritonFusedMoE` was selected in every run. torch is identical on both arms (2.12.0a0, NGC 26.05), so Triton is the only variable: base 3.6.0 + vendored v3.6.0 vs PR 3.7.0 + vendored v3.7.0.

| concurrency | 3.6.0 tok/s | 3.7.0 tok/s | Δ |
|---:|---:|---:|---:|
| 1 | 215.0 | 255.7 | **+18.9%** |
| 8 | 1270.9 | 1462.8 | +15.1% |
| 32 | 3431.2 | 3939.8 | +14.8% |
| 64 | 5331.2 | 5969.3 | +12.0% |
| 128 | 8492.4 | 9381.6 | +10.5% |
| 256 | 12320.3 | 11682.0 | **−5.2%** |

So the upgrade is not perf-neutral — it's concurrency-dependent, and your prediction here describes the shape exactly. 3.7 wins +10-19% through concurrency 128; at 256 the extra `_reduce_forward` launch per MoE layer outweighs the kernel gains by ~5%. Crossover sits between 128 and 256, which makes this short-circuit the highest-value item on the deferred perf list.

Controls: conc=1 re-run with the arm order reversed reproduced to within 0.1% (4763.5/4764.3 ms base, 4004.3/4002.0 ms PR), so it isn't a page-cache or ordering artifact. conc=256 was run three times; two agree at −5.2%/−5.0%, and the outlier (−9.1%) came from a noisier node where *both* arms were ~8% slower.

Caveat: TP4 rather than the TP8 CI shape — the 8-GPU partition was unschedulable. Happy to re-run at TP8 if you want the CI-matching numbers.

File: security_scanning/pyproject.toml [hunk 6]
```diff
@@ -65,7 +65,7 @@ dependencies = [
     "ninja (>=1.13.0,<2.0.0)",
     "blake3 (>=1.0.9,<2.0.0)",
     "soundfile (>=0.14.0,<0.15.0)",
-    "triton (==3.6.0)",
+    "triton (==3.7.0)",
```
Comment (BowenFu) [inline security_scanning/pyproject.toml:68]: This updates the scanner input to Triton 3.7.0, but the current PR head leaves `security_scanning/poetry.lock` at Triton 3.6.0. Please regenerate and include the lock from this dependency set; otherwise security scanning installs and validates a different runtime than the product and vendored-kernel migration.

**Comment by HandongLi-01** [comment]
@yuanjingx87 thanks for catching this — I don't think the regen is needed here, but tell me if you disagree.

All CI images build with `TORCH_INSTALL_TYPE=skip`, and `skip` is a no-op in `install_pytorch.sh:77`:
- `docker/Makefile:45` — `TORCH_INSTALL_TYPE ?= skip`
- `jenkins/BuildDockerImage.groovy:493` — `torchInstallType: "skip"`, the only value in the CI image matrix (no variant uses `pypi` or `src_*`)

So `TORCH_VERSION` isn't exercised when building CI images. The `2.11.0` → `2.12.0` edit only affects the `pypi` / `src_*` paths, keeping that script consistent with `requirements.txt`'s `torch>=2.12.0a0`; it doesn't change any image's contents.

Forcing a non-skip install would also be a downgrade in fidelity: `install_from_pypi()` uninstalls torch and installs public `torch==2.12.0`, replacing the NGC build the images ship (`2.12.0a0+...nv26.05`).

I did try the regen first — `BuildDockerImages` #165 failed at `make rockylinux8_push (x86_64)` on a missing UCX/DOCA header (`doca_gpunetio_verbs_def.h`). That message also appears in #159/#162/#164, which passed, so it looks pre-existing rather than something this PR introduced.

OK to skip the regen + `current_image_tags.properties` update? Happy to run it if you'd still prefer.

**Comment by HandongLi-01** [comment]
@brnguyen2 on Hopper/Blackwell coverage:

`TritonFusedMoE` is hard-gated to SM90 (`fused_moe_triton.py:1571-1576` returns a skip reason on any other SM), so there's no Blackwell integration path for this backend — hence both TRITON-backend GPT-OSS tests being `@skip_no_hopper`. The Blackwell layout branch you spotted is in the AutoDeploy path (`mxfp4_moe.py:62,76`, unswizzled under `cuda_capability_geq(10)`), which is opt-in there — AutoDeploy defaults to the `trtllm` backend on SM>=100 unless `backend="triton"` is set (`fused_moe_mxfp4.py:35-41`).

Coverage today:
- Hopper — `unittest/_torch/modules/fused_moe` (`l0_h100.yml`) + the two integration tests
- Blackwell — `unittest/_torch/modules/fused_moe` (`l0_b200.yml`/`l0_b300.yml`) and `unittest/auto_deploy/singlegpu/custom_ops/moe` (`l0_b200.yml`)
- H20 keeps the strided fallback (`is_swizzling_supported()`, nvbugs/5446119)

So Blackwell is unit-test only for the layout code. An integration-level Blackwell case would have to go via AutoDeploy with explicit `backend="triton"`.

File: security_scanning/pyproject.toml (hunk 6 above)
Comment (HandongLi-01) [reply security_scanning/pyproject.toml:68]: This file will be regenerated in CI and has conflicts from time to time so it is safe to leave this as it is.

File: requirements.txt [hunk 7]
```diff
@@ -22,10 +22,12 @@ h5py==3.12.1
...
 # https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel-26-05.html#rel-26-05 uses 2.12.0a0.
-torch>=2.11.0,<=2.13.0a0
+# The floor admits the NGC alpha builds (2.12.0a0+...); torch<2.12 pins triton==3.6.0,
+# which conflicts with the triton pin below.
+torch>=2.12.0a0,<=2.13.0a0
 torchvision
 # https://docs.nvidia.com/deeplearning/frameworks/pytorch-release-notes/rel-26-05.html#rel-26-05 uses 2.30.4
-# torch 2.11.0+cu130 depends on nvidia-nccl-cu13==2.28.9
+# torch 2.12.0+cu130 uses NCCL 2.29.7; NGC 26.05 uses NCCL 2.30.4.
 nvidia-nccl-cu13>=2.28.9,<=2.30.4
```
Comment (chzblych) [inline requirements.txt]:
```suggestion
nvidia-nccl-cu13>=2.29.7,<=2.30.4
```
